File: www/src/utils/filterFields.ts
import type { FilterOption } from './filterOptions'
import type { PossibleTarget, PossibleArtifact } from '../types/bridge'
import {
  getMissionFilterValueOptions,
  getTargetFilterOptions,
  getDropFilterOptions,
  getArtifactRarityFilterOptions,
  getArtifactSpecTypeFilterOptions,
  getArtifactTierFilterOptions,
  getArtifactNameFilterOptions,
} from './filterOptions'

export interface FilterFieldCtx {
  possibleTargets: PossibleTarget[]
  artifactConfigs: PossibleArtifact[]
  maxQuality: number
}

export type FilterValueKind = 'select' | 'modal' | 'date' | 'bool' | 'number'

export interface FilterOp {
  value: string
  label: string
}

export interface FilterFieldDef {
  key: string
  label: string
  scope: 'mission' | 'artifact'
  valueKind: FilterValueKind
  ops: FilterOp[]
  /** For 'select' and 'modal' kinds: builds the option list. */
  optionsSource?: (ctx: FilterFieldCtx) => FilterOption[]
}

const EQUALITY_OPS: FilterOp[] = [
  { value: '=', label: 'is' },
  { value: '!=', label: 'is not' },
]

const COMPARISON_OPS: FilterOp[] = [
  { value: '=', label: 'is' },
  { value: '!=', label: 'is not' },
  { value: '>', label: 'greater than' },
  { value: '<', label: 'less than' },
  { value: '>=', label: 'at least' },
  { value: '<=', label: 'at most' },
]

const DATE_OPS: FilterOp[] = [
  { value: '=', label: 'on' },
  { value: '<', label: 'before' },
  { value: '>', label: 'after' },
  { value: '<=', label: 'on or before' },
  { value: '>=', label: 'on or after' },
]

const DROPS_OPS: FilterOp[] = [
  { value: 'c', label: 'contains' },
  { value: 'dnc', label: 'does not contain' },
]

const BOOL_OPS: FilterOp[] = [
  { value: 'true', label: 'True' },
  { value: 'false', label: 'False' },
]

export const REPORT_FILTER_FIELDS: FilterFieldDef[] = [
  {
    key: 'ship',
    label: 'Ship',
    scope: 'mission',
    valueKind: 'select',
    ops: COMPARISON_OPS,
    optionsSource: () => getMissionFilterValueOptions('ship'),
  },
  {
    key: 'duration',
    label: 'Duration',
    scope: 'mission',
    valueKind: 'select',
    ops: COMPARISON_OPS,
    optionsSource: () => getMissionFilterValueOptions('duration'),
  },
  {
    key: 'level',
    label: 'Level',
    scope: 'mission',
    valueKind: 'select',
    ops: COMPARISON_OPS,
    optionsSource: () => getMissionFilterValueOptions('level'),
  },
  {
    key: 'target',
    label: 'Target',
    scope: 'mission',
    valueKind: 'modal',
    ops: EQUALITY_OPS,
    optionsSource: ctx => getTargetFilterOptions(ctx.possibleTargets),
  },
  {
    key: 'type',
    label: 'Mission Type',
    scope: 'mission',
    valueKind: 'select',
    ops: EQUALITY_OPS,
    optionsSource: () => getMissionFilterValueOptions('type'),
  },
  {
    key: 'launchDT',
    label: 'Launch Date',
    scope: 'mission',
    valueKind: 'date',
    ops: DATE_OPS,
  },
  {
    key: 'returnDT',
    label: 'Return Date',
    scope: 'mission',
    valueKind: 'date',
    ops: DATE_OPS,
  },
  {
    key: 'dubcap',
    label: 'Dub cap',
    scope: 'mission',
    valueKind: 'bool',
    ops: BOOL_OPS,
  },
  {
    key: 'buggedcap',
    label: 'Bugged cap',
    scope: 'mission',
    valueKind: 'bool',
    ops: BOOL_OPS,
  },
  {
    key: 'drops',
    label: 'Drops',
    scope: 'mission',
    valueKind: 'modal',
    ops: DROPS_OPS,
    optionsSource: ctx => getDropFilterOptions(ctx.artifactConfigs, ctx.maxQuality, true),
  },
  {
    key: 'artifact_name',
    label: 'Name',
    scope: 'artifact',
    valueKind: 'modal',
    ops: EQUALITY_OPS,
    optionsSource: ctx => getArtifactNameFilterOptions(ctx.artifactConfigs),
  },
  {
    key: 'artifact_rarity',
    label: 'Rarity',
    scope: 'artifact',
    valueKind: 'select',
    ops: COMPARISON_OPS,
    optionsSource: () => getArtifactRarityFilterOptions(),
  },
  {
    key: 'artifact_tier',
    label: 'Tier',
    scope: 'artifact',
    valueKind: 'select',
    ops: COMPARISON_OPS,
    optionsSource: ctx => getArtifactTierFilterOptions(ctx.artifactConfigs),
  },
  {
    key: 'artifact_spec_type',
    label: 'Spec Type',
    scope: 'artifact',
    valueKind: 'select',
    ops: EQUALITY_OPS,
    optionsSource: () => getArtifactSpecTypeFilterOptions(),
  },
  {
    key: 'artifact_quality',
    label: 'Quality',
    scope: 'artifact',
    valueKind: 'number',
    ops: COMPARISON_OPS,
  },
]

/** Returns the field definition for a given key, or undefined if no field matches. */
export function getReportField(key: string): FilterFieldDef | undefined {
  return REPORT_FILTER_FIELDS.find(f => f.key === key)
}

/**
 * Returns the default operator value for a field. Bool fields default to 'true' and drops defaults
 * to 'c'; every other field uses its first operator.
 */
export function defaultOpForField(def: FilterFieldDef): string {
  if (def.valueKind === 'bool') return 'true'
  if (def.key === 'drops') return 'c'
  return def.ops[0]?.value ?? ''
}

/** Returns the report filter fields scoped to missions. */
export function reportMissionFields(): FilterFieldDef[] {
  return REPORT_FILTER_FIELDS.filter(f => f.scope === 'mission')
}

/** Returns the report filter fields scoped to artifacts. */
export function reportArtifactFields(): FilterFieldDef[] {
  return REPORT_FILTER_FIELDS.filter(f => f.scope === 'artifact')
}

File: src/EggLedger.Web/Missions/FilterFields.cs
using EggLedger.Domain.MissionQuery;

namespace EggLedger.Web.Missions;

public sealed class FilterFieldCtx {
    public IReadOnlyList<PossibleTarget> PossibleTargets { get; init; } = Array.Empty<PossibleTarget>();
    public IReadOnlyList<PossibleArtifact> ArtifactConfigs { get; init; } = Array.Empty<PossibleArtifact>();
    public double MaxQuality { get; init; }
}

public sealed class FilterFieldDef {
    public string Key { get; init; } = "";
    public string Label { get; init; } = "";

    /// <summary>"mission" or "artifact".</summary>
    public string Scope { get; init; } = "mission";
    public FilterValueKind ValueKind { get; init; }
    public IReadOnlyList<FilterOp> Ops { get; init; } = Array.Empty<FilterOp>();

    /// <summary>For select/modal kinds: builds the option list. Null otherwise.</summary>
    public Func<FilterFieldCtx, List<FilterOption>>? OptionsSource { get; init; }
}

/// <summary>Static filter field/operator definitions, golden-matched to filterFields.ts (field set, operator lists, default-operator rules, scope partitions).</summary>
public static class FilterFields {
    public static readonly IReadOnlyList<FilterOp> EqualityOps = new[]
    {
        new FilterOp("=", "is"),
        new FilterOp("!=", "is not"),
    };

    public static readonly IReadOnlyList<FilterOp> ComparisonOps = new[]
    {
        new FilterOp("=", "is"),
        new FilterOp("!=", "is not"),
        new FilterOp(">", "greater than"),
        new FilterOp("<", "less than"),
        new FilterOp(">=", "at least"),
        new FilterOp("<=", "at most"),
    };

    public static readonly IReadOnlyList<FilterOp> DateOps = new[]
    {
        new FilterOp("=", "on"),
        new FilterOp("<", "before"),
        new FilterOp(">", "after"),
        new FilterOp("<=", "on or before"),
        new FilterOp(">=", "on or after"),
    };

    /// <summary>Date operators for the Mission Data tab filter bar. "on" must map to "d=" (day-equality), since a date "=" reference-compares and never matches (see MissionFilterMatcher).</summary>
    public static readonly IReadOnlyList<FilterOp> MissionDateOps = new[]
    {
        new FilterOp("d=", "on"),
        new FilterOp("<", "before"),
        new FilterOp(">", "after"),
    };

    /// <summary>Operators the Mission Data filter bar presents for a field. Date fields use MissionDateOps; everything else uses the field's own Ops.</summary>
    public static IReadOnlyList<FilterOp> MissionBarOpsFor(FilterFieldDef def) =>
        def.ValueKind == FilterValueKind.Date ? MissionDateOps : def.Ops;

    public static readonly IReadOnlyList<FilterOp> DropsOps = new[]
    {
        new FilterOp("c", "contains"),
        new FilterOp("dnc", "does not contain"),
    };

    public static readonly IReadOnlyList<FilterOp> BoolOps = new[]
    {
        new FilterOp("true", "True"),
        new FilterOp("false", "False"),
    };

    /// <summary>All filter fields, mission then artifact scope.</summary>
    public static readonly IReadOnlyList<FilterFieldDef> ReportFilterFields = new[]
    {
        new FilterFieldDef
        {
            Key = "ship", Label = "Ship", Scope = "mission", ValueKind = FilterValueKind.Select,
            Ops = ComparisonOps, OptionsSource = _ => FilterOptions.GetMissionFilterValueOptions("ship"),
        },
        new FilterFieldDef
        {
            Key = "duration", Label = "Duration", Scope = "mission", ValueKind = FilterValueKind.Select,
            Ops = ComparisonOps, OptionsSource = _ => FilterOptions.GetMissionFilterValueOptions("duration"),
        },
        new FilterFieldDef
        {
            Key = "level", Label = "Level", Scope = "mission", ValueKind = FilterValueKind.Select,
            Ops = ComparisonOps, OptionsSource = _ => FilterOptions.GetMissionFilterValueOptions("level"),
        },
        new FilterFieldDef
        {
            Key = "target", Label = "Target", Scope = "mission", ValueKind = FilterValueKind.Modal,
            Ops = EqualityOps, OptionsSource = ctx => FilterOptions.GetTargetFilterOptions(ctx.PossibleTargets),
        },
        new FilterFieldDef
        {
            Key = "type", Label = "Mission Type", Scope = "mission", ValueKind = FilterValueKind.Select,
            Ops = EqualityOps, OptionsSource = _ => FilterOptions.GetMissionFilterValueOptions("type"),
        },
        new FilterFieldDef
        {
            Key = "launchDT", Label = "Launch Date", Scope = "mission", ValueKind = FilterValueKind.Date,
            Ops = DateOps,
        },
        new FilterFieldDef
        {
            Key = "returnDT", Label = "Return Date", Scope = "mission", ValueKind = FilterValueKind.Date,
            Ops = DateOps,
        },
        new FilterFieldDef
        {
            Key = "dubcap", Label = "Dub cap", Scope = "mission", ValueKind = FilterValueKind.Bool,
            Ops = BoolOps,
        },
        new FilterFieldDef
        {
            Key = "buggedcap", Label = "Bugged cap", Scope = "mission", ValueKind = FilterValueKind.Bool,
            Ops = BoolOps,
        },
        new FilterFieldDef
        {
            Key = "drops", Label = "Drops", Scope = "mission", ValueKind = FilterValueKind.Modal,
            Ops = DropsOps,
            OptionsSource = ctx => FilterOptions.GetDropFilterOptions(ctx.ArtifactConfigs, ctx.MaxQuality, true),
        },
        new FilterFieldDef
        {
            Key = "artifact_name", Label = "Name", Scope = "artifact", ValueKind = FilterValueKind.Modal,
            Ops = EqualityOps, OptionsSource = ctx => FilterOptions.GetArtifactNameFilterOptions(ctx.ArtifactConfigs),
        },
        new FilterFieldDef
        {
            Key = "artifact_rarity", Label = "Rarity", Scope = "artifact", ValueKind = FilterValueKind.Select,
            Ops = ComparisonOps, OptionsSource = _ => FilterOptions.GetArtifactRarityFilterOptions(),
        },
        new FilterFieldDef
        {
            Key = "artifact_tier", Label = "Tier", Scope = "artifact", ValueKind = FilterValueKind.Select,
            Ops = ComparisonOps, OptionsSource = ctx => FilterOptions.GetArtifactTierFilterOptions(ctx.ArtifactConfigs),
        },
        new FilterFieldDef
        {
            Key = "artifact_spec_type", Label = "Spec Type", Scope = "artifact", ValueKind = FilterValueKind.Select,
            Ops = EqualityOps, OptionsSource = _ => FilterOptions.GetArtifactSpecTypeFilterOptions(),
        },
        new FilterFieldDef
        {
            Key = "artifact_quality", Label = "Quality", Scope = "artifact", ValueKind = FilterValueKind.Number,
            Ops = ComparisonOps,
        },
    };

    /// <summary>Field definition for a key, or null.</summary>
    public static FilterFieldDef? GetReportField(string key) {
        foreach (var f in ReportFilterFields) {
            if (f.Key == key) {
                return f;
            }
        }
        return null;
    }

    /// <summary>Default operator for a field in the Mission Data bar: bool -> "true", drops -> "c", date -> "d=", else first operator.</summary>
    public static string DefaultOpForField(FilterFieldDef def) {
        if (def.ValueKind == FilterValueKind.Bool) {
            return "true";
        }
        if (def.Key == "drops") {
            return "c";
        }
        var ops = MissionBarOpsFor(def);
        return ops.Count > 0 ? ops[0].Value : "";
    }

    /// <summary>Mission-scoped fields.</summary>
    public static List<FilterFieldDef> ReportMissionFields() {
        var result = new List<FilterFieldDef>();
        foreach (var f in ReportFilterFields) {
            if (f.Scope == "mission") {
                result.Add(f);
            }
        }
        return result;
    }

    /// <summary>Artifact-scoped fields.</summary>
    public static List<FilterFieldDef> ReportArtifactFields() {
        var result = new List<FilterFieldDef>();
        foreach (var f in ReportFilterFields) {
            if (f.Scope == "artifact") {
                result.Add(f);
            }
        }
        return result;
    }
}
